### ISSUE TITLE
Pause causes old .seg files to no longer be removed resulting in ever increasing storage.

### ISSUE THREAD

**Issue by FXB78** [xbmc/inputstream.ffmpegdirect#102]
Where we have a 'Maximum timeshift buffer length' of e.g 1 hour, the initial .seg files will start to be slowly deleted once the hour limit is reached (so 0.seg, then 1.seg, then 2.seg etc are sequentially deleted). This all works fine until the stream is paused, so lets say obsolete .seg files are deleted up to 32.seg, at which time I pause the stream for 30 secs, then resume watching. I would expect 32.seg, 33.seg etc to continue to be removed even when the stream is paused as they are still being replaced by current .seg files. However they are now left on storage and even after playback resumes no further .seg files are ever removed (until playback is completely stopped).

So in Summary once the Max buffer length is reached a Pause of the stream will cause obsolete files to no longer be removed, and therefore my set limit of one hour no longer applies & could reach two, three or more hours simply because I paused the stream at some point.

In my view old .seg files should continue to be removed even when paused thus honouring the buffer limit set by the user.

**Comment by phunkyfish** [comment]
If old seg files are removed how will you resume from pause?

**Comment by FXB78** [comment]
I think that if you leave it paused past the buffer limit you have set it should just time out. That seems to be a more reasonable solution than never removing them again even once playback has begun.

If I set a max buffer value of 1 hour and leave it for over an hour I'm not expecting to resume the stream.

In case you misunderstood I'm not saying remove ALL seg files, just those which are older than the buffer time you have set.

For example a use case of mine today:

Storage space remaining = 3.5GB
1 hour stream = 2GB approx.

Play stream for an hour, Storage space now = 1.5GB free
Pause the stream for 30 seconds and resume, storage still 1.5GB free
Come back an hour later, storage space now = 0GB as following pause no seg files have been deleted & the buffer length is now >2 hours, even though I set Max 1 hour in settings.

**Comment by phunkyfish** [comment]
This is not possible. The VideoPlayer in Kodi is paused and holds that state so when resumed it will assume the stream is at one point where it fact it will be way ahead. The add-on needs to be in sync with VP and it's not possible to change the point it's paused at as far as I recall.

I think the only option might be to stop the stream instead of going over the limit. I can look at that and see if it's possible. I may have already tried to do this and not succeeded. But can look again when I have some free time.

EDIT: I see the problem. So VP calls DemuxRead() on the inputstream to read more packets of data. But when the stream is paused it stops reading packets. So there is no way to tell it stop the stream because to do so it would need to read a packet with an EOF or return `nullptr` to signify an error. I'm not sure there is an easy way around this.

**Comment by phunkyfish** [comment]
I would like to point out that help text for the timeshift buffer limit clearly states that the buffer will grow indefinitely when paused.

**Comment by phunkyfish** [comment]
What happens if you pause so it goes over the limit then resume and seek to current live time? Does it correctly remove the extra .seg files?

**Comment by FXB78** [comment]
> I would like to point out that help text for the timeshift buffer limit clearly states that the buffer will grow indefinitely when paused.

Yes this is fine, my issue was more that upon resuming streams it didn't clear the old files via RemoveOldestInMemoryAndOnDiskSegments.


> What happens if you pause so it goes over the limit then resume and seek to current live time? Does it correctly remove the extra .seg files?

This was the issue, on resuming the stream the old seg files (>1 hour) were not being removed. I've been testing it and out of 4 tests I could only reproduce it once, the other 3 times it correctly removed the old seg files on resuming the stream as expected. Unfortunately the time I reproduced it Debugging wasn't enabled so the log is limited.

The Log shows 'messy timestamps' so I'm wondering if this is causing the issue, the relevant part of the log is here https://paste.kodi.tv/uyicalusey.kodi

And these were the files with a 15 min limit, showing 20 mins & increasing all the time after playback was resumed at 21:03:

![image](https://user-images.githubusercontent.com/80767985/116618362-dd3ca780-a936-11eb-8a69-2c095799acbf.png)

**Comment by phunkyfish** [comment]
What do you mean by resuming the stream? Pressing play or seeking to live playback (i.e. now)?

**Comment by FXB78** [comment]
> What do you mean by resuming the stream? Pressing play or seeking to live playback (i.e. now)?

Just pressing Play to resume from where it was paused.

I managed to recreate with Debug Logging enabled, as far as I can tell it seems to invoke RemoveOldestInMemoryAndOnDiskSegments but from the wrong seg file, in this example whilst paused it starts with ID:23 but I have older seg files which don't get removed. I'm not sure if 23, 24 actually got removed, I suspect not but didn't actually check each one at the time. These are the Files after playback was resumed at 22:12:

![image](https://user-images.githubusercontent.com/80767985/116624022-1711ac00-a93f-11eb-8db0-054531cb10ed.png)

and this is the relevant part of the log from where it was paused -> 
[timeshift.txt](https://github.com/xbmc/inputstream.ffmpegdirect/files/6401885/timeshift.txt)

**Comment by phunkyfish** [comment]
The moment you press resume 0 files should be removed. Only once there are enough files that are no longer required, I.e. you played at least segment length minutes will a file be removed.

So when you press play at 12:12 that doesn’t mean all files from before that time are removed. The files from when pause was pressed are still needed.

This is the critical piece in `RemoveOldestInMemoryAndOnDiskSegments`:

```cpp
  if (m_enableOnDiskSegmentLimit && !m_paused &&
      m_segmentTotalCount > m_maxOnDiskSegments &&
      m_currentDemuxTimeIndex > m_minOnDiskSeekTimeIndex)
```

Because the playback time is when it was paused (i.e. `m_currentDemuxTimeIndex < m_minOnDiskSeekTimeIndex`) it can't pass the condition that removes the segment files.

**Comment by FXB78** [comment]
Yes I understand that at 22:12 the files will still be there, so I waited 2 minutes & checked the files in the screenshots at 22:14, I would expect the oldest segments to be removed at this time as it had create several additional new files.

I'm not sure we're exactly on the same page here, so these are the steps I took to reproduce the issue:

Play Stream with 15 minute Max Buffer Limit set
Wait 16/17 minutes (Buffer is at 15 minutes & 00-07.seg files have been correctly removed sequentially).
Press Pause
Observe in File Manager oldest seg file (08.seg)
Leave Paused for 2-3 minutes
Press Play to resume Playback
Wait 2 minutes
Check File Manager & oldest seg file is still 08.seg
New seg files 103-110 have been created while oldest files (08.seg, 09.seg, 10.seg) have not been removed.
Timeshift buffer in Kodi is now >15 minute Limit due to seg files outside the limit not being deleted as they were being prior to me pressing Pause.

For whatever reason the Pause & Resume is causing the oldest seg files to remain indefinitely instead of being deleted.

**Comment by phunkyfish** [comment]
Ok, this time do the same test but when you press pause 1) note the lowest and high segment number. Wait a few minutes and press play. 2) Note the numbers again.

Now wait 10 minutes. 3) Note them again.

Now seek to live and 4) note them again.

Report the 4 sets of values.

1) press pause
2) wait 3 mins then press play
3) wait 10 minutes 
4) seek to live

**Comment by FXB78** [comment]
OK I've performed those tests, results as below:

![TS Tests](https://user-images.githubusercontent.com/80767985/116699070-115aab80-a9bd-11eb-99e8-e98391859f79.png)

**Comment by FXB78** [comment]
I just had a thought about this, I was using Unpause Jumpback, which I think could be causing problems as it jumps back 3 seconds. I'll disable this addon & try some more tests.

**Comment by FXB78** [comment]
> I just had a thought about this, I was using Unpause Jumpback, which I think could be causing problems as it jumps back 3 seconds. I'll disable this addon & try some more tests.

So I disabled Unpause Jumpback and thought it had solved it but it has just started happening again. It's a strange issue as it doesn't happen every time so it's difficult to pinpoint the root-cause.

**Comment by phunkyfish** [comment]
Hmmm, ok. If we can pinpoint the cause or what is different when it happens I can fix it.

Only thing is I have tried 20/30 times and I can’t reproduce following your steps.

**Comment by FXB78** [comment]
Yes it's definitely a strange issue, looking at the log I posted it doesn't seem to make sense. It appears to be trying to remove segments whilst paused, which of course shouldn't be happening:

22:09:10.588 HandleKey: blue (0xe6) pressed, action is Pause
22:09:10.624 CDVDAudio::Pause - pausing audio stream
22:09:14.645 inputstream.ffmpegdirect: AddPacket - Writing new segment - seconds: 1020, last seg seconds: 1008, last seg packet count: 817, new seg index: 85, pts 1020000000.00, dts: 1020000000.00, pts sec: 1020, dts sec: 1020
22:09:14.645 inputstream.ffmpegdirect: RemoveOldestInMemoryAndOnDiskSegments - Removed oldest in memory segment with ID: 23
22:09:26.664 inputstream.ffmpegdirect: AddPacket - Writing new segment - seconds: 1032, last seg seconds: 1020, last seg packet count: 817, new seg index: 86, pts 1032018100.00, dts: 1032018100.00, pts sec: 1032, dts sec: 1032
22:09:26.665 inputstream.ffmpegdirect: RemoveOldestInMemoryAndOnDiskSegments - Removed oldest in memory segment with ID: 24
......
22:11:26.916 inputstream.ffmpegdirect: RemoveOldestInMemoryAndOnDiskSegments - Removed oldest in memory segment with ID: 34
2021-04-29 inputstream.ffmpegdirect: TimeshiftSegment - Segment ID: 96, Segment Filename: 1743746610-00000096.seg
......
22:12:18.379 CDVDAudio::Resume - resume audio stream
22:12:18.382 ActiveAE - start sync of audio stream
22:12:18.417 CVideoPlayerVideo::OutputPicture - ttd:355ms pts:981.260 Clock:980.904 Level:6


I'm wondering if something goes out of sync when this erroneous action is happening (causing it to no longer correctly delete seg files), and what is causing it in the first place.

**Comment by phunkyfish** [comment]
At this time are segments older than 23/24 still on the disk?

**Comment by FXB78** [comment]
> At this time are segments older than 23/24 still on the disk?

Yes, the oldest segment in that example was 08.seg which was also still there, as were 09,10,11 etc.

**Comment by phunkyfish** [comment]
It will always remove in memory segments because they are on disk. There are only so many it keeps in memory.

The question is why is it not deleting the on disk segments that are no longer needed. Let me add log statement for the values in the check and see if that gives us any clues.

Testbuilds: https://jenkins.kodi.tv/blue/organizations/jenkins/xbmc%2Finputstream.ffmpegdirect/detail/PR-103/1/artifacts

**Comment by FXB78** [comment]
> It will always remove in memory segments because they are on disk. There are only so many it keeps in memory.
> 
> The question is why is it not deleting the on disk segments that are no longer needed. Let me add log statement for the values in the check and see if that gives us any clues.

Ah yes of course, I didn't realise there were both On Disk & Memory segments, that makes sense.

The box I am seeing the issue on is running CoreELEC, is there a test build I can install on CoreELEC or would I need to boot into Android? It would be ideal if I could install on CoreELEC but if it's not possible I can test in Android.

**Comment by phunkyfish** [comment]
Not possible. Only the standard Kodi platforms can be built by Jenkins.

**Comment by phunkyfish** [comment]
Do you have a windows or mac machine you can try this on for testing?

**Comment by FXB78** [comment]
> Do you have a windows or mac machine you can try this on for testing?

Yes, I've been testing on both Windows & Android but have been unable to reproduce the issue. So I'm wondering if it's a CoreELEC/Linux specific issue, but I've no idea what it could be unless there is something happening in the CoreELEC build process.

As a work-around I can remove the limit & move the timeshift to external storage, it's just strange that it consistently happens but only on CoreELEC.

**Comment by phunkyfish** [comment]
Ah, if it’s only on coreelec it could be a 32bit time problem. I remember finding issues previously in some PVR add-ons.

Makes it tricky to debug though as I don’t have a comparable test system.

**Comment by FXB78** [comment]
> Ah, if it’s only on coreelec it could be a 32bit time problem. I remember finding issues previously in some PVR add-ons.
> 
> Makes it tricky to debug though as I don’t have a comparable test system.

Yes it certainly seems to be specific to CoreELEC and not a wider issue affecting all Kodi versions.

In the absence of any way to progress further I'm happy to close the issue & just work-around it by removing the limit and using external storage.

Thanks for all your help in looking into the issue.

**Comment by phunkyfish** [comment]
Maybe the guys over at CoreElec can add that log statement?

It's here: https://github.com/xbmc/inputstream.ffmpegdirect/pull/103

If we can get that in and running I'm pretty sure we can figure it out.

Maybe @arthur-liberman can help?

**Comment by arthur-liberman** [comment]
This CE build is based on our nightly build and includes the patch above.
`inputstream.ffmpegdirect: 1.21.1-Matrix` and `kodi: #1a1c5028f24a992134e2a1745519914e303eee4c`
https://test.coreelec.org/arthur/CoreELEC-Amlogic-ng.arm-19.1-Matrix_devel_20210502230348.tar

**Comment by phunkyfish** [comment]
> This CE build is based on our nightly build and includes the patch above.
> `inputstream.ffmpegdirect: 1.21.1-Matrix` and `kodi: #1a1c5028f24a992134e2a1745519914e303eee4c`
> https://test.coreelec.org/arthur/CoreELEC-Amlogic-ng.arm-19.1-Matrix_devel_20210502230348.tar

Thanks!

**Comment by FXB78** [comment]
> This CE build is based on our nightly build and includes the patch above.
> `inputstream.ffmpegdirect: 1.21.1-Matrix` and `kodi: #1a1c5028f24a992134e2a1745519914e303eee4c`
> https://test.coreelec.org/arthur/CoreELEC-Amlogic-ng.arm-19.1-Matrix_devel_20210502230348.tar

Thanks Arthur, I'll update to this version & run a few more tests.

**Comment by arthur-liberman** [comment]
Note that the change only has some extra logging information. So try to reproduce this with debug mode enabled.

**Comment by phunkyfish** [comment]
> Note that the change only has some extra logging information. So try to reproduce this with debug mode enabled.

Yes, I reckon there is a 64/32bit conversion error and hopefully the logging will help track it down.

**Comment by phunkyfish** [comment]
@FXB78 

Please run a windows test alongside a CoreElec test at the same time. Being able to compare logs where the actions happen at roughly the same time should allow us to figure this out.

thanks!

**Comment by arthur-liberman** [comment]
I haven't realized that this is the addon part, not kodi part.
I'll build the addon after work and upload it for @FXB78 to update and try.
Apologies.

**Comment by vpeter4** [comment]
Is it possible to provide step by step instructions how to repeat this issue? Like "start playing, wait 3 minutes, press pause, wait 10 minutes, ...". This way others can help too.

**Comment by phunkyfish** [comment]
Of course. @FXB78 if you could detail the steps for the test case that would be great.

**Comment by FXB78** [comment]
> Of course. @FXB78 if you could detail the steps for the test case that would be great.

Yes of course, it's quite simple to replicate but is an intermittent issue:

Set Buffer time length to 0.25 (15 min)
Play
Wait >15 minutes until buffer is full & .seg files start to be removed
Pause for 3/4 minutes
Note down oldest .seg file (from timeshift folder, i.e 08.seg)
Resume Playback by pressing Play
Wait 2 minutes & check oldest .seg file
Oldest .seg files should continue to be removed in normal operation, in the case of this bug the oldest file will remain as the one we noted down earlier (i.e 08.seg).

**Comment by vpeter4** [comment]
I did like written above but few seconds after resume oldest .seq files were deleted. Which means I can't replicate this issue. Did same steps twice after reboot.
Using CoreELEC 19.1 rc4, pvr.iptvsimple 7.6.1.1, inputstream.ffmpegdirect 1.21.1.1.

**Comment by FXB78** [comment]
After several attempts I've managed to recreate this issue in CoreELEC, unfortunately it wasn't practical to test concurrently with Windows.

This is an excerpt from the log just before Pause:
[timeshift03051603.txt](https://github.com/xbmc/inputstream.ffmpegdirect/files/6415911/timeshift03051603.txt)

And these were the files as shown via SSH:

![timeshift0305](https://user-images.githubusercontent.com/80767985/116898457-4d487780-ac2e-11eb-8d2f-4b06fb962c2b.png)

**Comment by FXB78** [comment]
I left it running longer and it lets me successfully navigate back over an hour despite the limit being set to 15 mins (ignore the artifacts, the image on screen was perfect):

![screenshot00000](https://user-images.githubusercontent.com/80767985/116900011-2e4ae500-ac30-11eb-8728-b828755071c0.png)

**Comment by vpeter4** [comment]
From what you wrote this doesn't happen ALWAYS? In this case I will do some more tries tomorrow.

And you missed the log AFTER resume. This part is important.

**Comment by FXB78** [comment]
> From what you wrote this doesn't happen ALWAYS? In this case I will do some more tries tomorrow.
> 
> And you missed the log AFTER resume. This part is important.

Yes it doesn't happen every time, perhaps 20% of the time. In the log the Pause is at Line 104, whilst Resume is just after at line 149.

**Comment by arthur-liberman** [comment]
Here's the addon with the patch included.
https://test.coreelec.org/arthur/inputstream.ffmpegdirect-1.21.1.1.zip

Make sure that you restart Kodi after updating, just to make sure it updated correctly.

**Comment by phunkyfish** [comment]
> Here's the addon with the patch included.
> https://test.coreelec.org/arthur/inputstream.ffmpegdirect-1.21.1.1.zip

Thanks. @FXB78 can you repro with this and post the log?

**Comment by FXB78** [comment]
> > Here's the addon with the patch included.
> > https://test.coreelec.org/arthur/inputstream.ffmpegdirect-1.21.1.1.zip
> 
> Thanks. @FXB78 can you repro with this and post the log?

This is an excerpt using the patched addon 
[ts1750.txt](https://github.com/xbmc/inputstream.ffmpegdirect/files/6416442/ts1750.txt)


in the log we see on line 26 "RemoveOldestInMemoryAndOnDiskSegments - Removed oldest on disk segment with ID: 43"
Paused at line 44 "HandleKey: blue (0xe6) pressed, action is Pause"
Resume at line 154 "HandleKey: homepage (0xea) pressed, action is PlayPause"
Then at line 1185 we see "RemoveOldestInMemoryAndOnDiskSegments - Removed oldest on disk segment CHECK enabled: 1, paused: 1" Which I think seems to indicate the system thinks the stream is still paused & thus doesn't remove the old seg files?

And these are the files with 44.seg etc never being removed:
![ts1750](https://user-images.githubusercontent.com/80767985/116908650-1036b200-ac3b-11eb-9687-8eab19ebc11a.png)

**Comment by phunkyfish** [comment]
Yes. That’s it. It looks like the unpause is not captured. Now we just need to find out why.

**Comment by phunkyfish** [comment]
Please uninstall unpause jumpback completely. SetPaused is only called on pause and is missing for unpause. This likely means this other addon is causing the issue.

So I don’t think anyone else will be able to repro this.

**Comment by FXB78** [comment]
> Please uninstall unpause jumpback completely. SetPaused is only called on pause and is missing for unpause. This likely means this other addon is causing the issue.
> 
> So I don’t think anyone else will be able to repro this.

Yes I think you're right, and this would tie-in with why I could only reproduce it on the one device which had Unpause Jumpback installed. So it's not a CoreELEC issue, it's just that CoreELEC was the only OS that had the addon installed.

The Unpause Jumpback addon was what I had earlier thought might be the culprit, but after seemingly disabling it the issue remained. I have been running more tests and playing about with Unpause Jumpback & have noticed some very odd caching behaviour which would've most likely meant it wasn't fully disabled until as you suggested I fully uninstalled it. After this I have not seen the issue again.

So I'm happy to close this now on the basis that I can no longer reproduce with Unpause Jumpback removed.

Thanks to everyone involved for giving their time & input, it's really appreciated and I'm glad we finally got to the bottom of it!

**Comment by FXB78** [comment]
Issue caused by conflicts with the Unpause Jumpback addon not calling SetPaused. Uninstalling Unpause Jumpback resolves the issue.
